lib/modulesync.rb

```diff
@@ -101,20 +102,18 @@ def self.manage_module(puppet_module, module_files, module_options, defaults, op
       raise unless options[:skip_broken]
     end
 
-    puts "Syncing #{puppet_module}"
     namespace, module_name = module_name(puppet_module, options[:namespace])
+    git_repo = "#{namespace}/#{module_name}"
```

Comment:
I don't think this is correct. IMHO you should introduce a new variable `namespace_root` that is the project root + namespace instead.

Reply:
Is this just a matter of taste, or a technical issue?

> new variable `namespace_root` that is the project root + namespace

What do you intend with "project root"? Not the repository? Can you make a concrete example?

P.S.: You can write, _"Replace this by ..."_. I promise, I won't be offended.

Reply:
> Is this just a matter of taste, or a technical issue?

I think it should be that way because the code that's called (`Git.pull`) does a `Dir.mkdir(project_root) unless Dir.exist?(project_root)`. Your version changes to rely on `git clone` to create parent directories. My git (2.14.4) does this, but I don't know if all git versions do this. If we rely on git to create parent directories, we might as well remove the `Dir.mkdir` there. This is me trying to figure out what's going on because I don't know the project that well.

> What do you intend with "project root"? Not the repository? Can you make a concrete example?

I meant `namespace_root = File.join(options[:project_root], namespace)` (though I'm not attached to the variable name).

> P.S.: You can write, "Replace this by ...". I promise, I won't be offended.

That sounds like I know what I'm doing :P

Reply:
I think this problem is artificial. When we look at the [Git.pull](https://github.com/voxpupuli/modulesync/blob/master/lib/modulesync/git.rb#L48-L52) code we see that the `Dir.mkdir` is needed. But not for the Git operation, but because we want to do Git things _in_ that directory.

We can still try to prove this is false and will fail. If we manage to come up with a test case that fails. I think that's the route that makes really sense. (I rely a lot on tests, especially for languages that I don't fully master. The existing ones have helped me to get as far as I am now.)